File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from collections import defaultdict
import math

def solution(fees, records):
    parking_info_map = defaultdict(ParkingInfo)

    for record in records:
        parking_data = record.split(" ")

        time = int(parking_data[0].split(":")[0]) * 60
        time += int(parking_data[0].split(":")[1])
        car_num = parking_data[1]
        breakdown = parking_data[2]

        if car_num not in parking_info_map:
            parking_info = ParkingInfo(time, car_num, breakdown)
            parking_info_map[car_num] = parking_info
        else:
            parking_info = parking_info_map[car_num]
            parking_info.parking_in_out(time, breakdown)

    sort_key = sorted(parking_info_map.keys())
    charge_list = [parking_info_map[key].total_cal(fees) for key in sort_key]

    return charge_list

class ParkingInfo:
    def __init__(self, time, car_num, status):
        self.time = time
        self.car_num = car_num
        self.status = status
        self.parking_time = 0

    def parking_in_out(self, time, check):
        if check == "IN":
            self.time = time
        else:
            self.parking_time += time - self.time
        self.status = check

    def total_cal(self, fees):
        basic_time, charge, unit_time, unit_charge = fees

        if self.status == "IN":
            self.parking_time += ((23 * 60) + 59) - self.time
            self.status = "OUT"

        if self.parking_time > basic_time:
            charge += math.ceil((self.parking_time - basic_time) / unit_time) * unit_charge

        return charge

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;
class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, ParkingInfo> parkingInfoMap = new HashMap<>();
        for (String record : records) {
            String[] parkingData = record.split(" ");

            int time = Integer.parseInt(parkingData[0].split(":")[0]) * 60;
            time += Integer.parseInt(parkingData[0].split(":")[1]);
            String carNum = parkingData[1];
            String breakDown = parkingData[2];

            if (!parkingInfoMap.containsKey(carNum)) {
                ParkingInfo parkingInfo = new ParkingInfo(time, carNum, breakDown);
                parkingInfoMap.put(carNum, parkingInfo);
            }
            else {
                ParkingInfo parkingInfo = parkingInfoMap.get(carNum);
                parkingInfo.parkingInOut(time,breakDown);
            }
        }

        List<String> sortKey = new ArrayList<>(parkingInfoMap.keySet());
        Collections.sort(sortKey);
        List<Integer> chargeList = new ArrayList<>();

        for (String key : sortKey) {
            chargeList.add(parkingInfoMap.get(key).totalCal(fees));
        }

        return chargeList.stream().mapToInt(Integer::intValue).toArray();
    }

    public static class ParkingInfo {
        int time;
        String carNum;
        String status = "NONE";
        int parkingTime = 0;

        public String getStatus() {
            return status;
        }

        public ParkingInfo(int time, String carNum, String status) {
            this.time = time;
            this.carNum = carNum;
            this.status = status;
        }

        public void parkingInOut(int time, String check) {
            if (check.equals("IN")) {
                this.time = time;
            } else {
                parkingTime += time - this.time;
            }
            status = check;
        }

        public int totalCal(int[] fees) {
            int basicTime = fees[0];
            int charge = fees[1];
            long unitTime = fees[2];
            int unitCharge = fees[3];

            if (Objects.equals(status, "IN")) {
                parkingTime += ((23*60)+59) - time;
                status = "OUT";
            }
            if (parkingTime > basicTime) {
                charge += Math.ceil((parkingTime - basicTime) / (unitTime + 0.0)) * unitCharge;
            }

            return charge;
        }
    }
}